dpcpp/matrix/ell_kernels.dp.cpp

```diff
@@ -57,12 +72,567 @@ namespace dpcpp {
+    const size_type stride, const size_type num_stored_elements_per_row,
+    acc::range<b_accessor> b, OutputValueType *__restrict__ c,
+    const size_type c_stride, sycl::nd_item<3> item_ct1,
+    UninitializedArray<OutputValueType,
+                       default_block_size / num_thread_per_worker> *storage)
+{
+    spmv_kernel<num_thread_per_worker, atomic>(
+        num_rows, num_worker_per_row, val, col, stride,
+        num_stored_elements_per_row, b, c, c_stride,
+        [](const OutputValueType &x, const OutputValueType &y) { return x; },
+        item_ct1, storage);
+}
+
+template <int num_thread_per_worker, bool atomic = false, typename b_accessor,
+          typename a_accessor, typename OutputValueType, typename IndexType>
+void spmv(dim3 grid, dim3 block, gko::size_type dynamic_shared_memory,
+          sycl::queue *stream, const size_type num_rows,
+          const int num_worker_per_row, acc::range<a_accessor> val,
+          const IndexType *col, const size_type stride,
+          const size_type num_stored_elements_per_row, acc::range<b_accessor> b,
+          OutputValueType *c, const size_type c_stride)
+{
+    stream->submit([&](sycl::handler &cgh) {
+        sycl::accessor<
+            UninitializedArray<OutputValueType,
+                               default_block_size / num_thread_per_worker>,
+            0, sycl::access_mode::read_write, sycl::access::target::local>
+            storage_acc_ct1(cgh);
+
+        cgh.parallel_for(sycl_nd_range(grid, block),
+                         [=](sycl::nd_item<3> item_ct1) {
+                             spmv<num_thread_per_worker, atomic>(
+                                 num_rows, num_worker_per_row, val, col, stride,
+                                 num_stored_elements_per_row, b, c, c_stride,
+                                 item_ct1, storage_acc_ct1.get_pointer().get());
+                         });
+    });
+}
+
+
+template <int num_thread_per_worker, bool atomic = false, typename b_accessor,
+          typename a_accessor, typename OutputValueType, typename IndexType>
+void spmv(
+    const size_type num_rows, const int num_worker_per_row,
+    acc::range<a_accessor> alpha, acc::range<a_accessor> val,
+    const IndexType *__restrict__ col, const size_type stride,
+    const size_type num_stored_elements_per_row, acc::range<b_accessor> b,
+    const OutputValueType *__restrict__ beta, OutputValueType *__restrict__ c,
+    const size_type c_stride, sycl::nd_item<3> item_ct1,
+    UninitializedArray<OutputValueType,
+                       default_block_size / num_thread_per_worker> *storage)
+{
+    const OutputValueType alpha_val = alpha(0);
+    const OutputValueType beta_val = beta[0];
+    // Because the atomic operation changes the values of c during computation,
+    // it can not do the right alpha * a * b + beta * c operation.
+    // Thus, the dpcpp kernel only computes alpha * a * b when it uses atomic
+    // operation.
+    if (atomic) {
```

Comment:
Maybe instead note that `c` needs to be scaled beforehand, otherwise this might lead to confusion. Also:
```suggestion
    // Because the atomic operation changes the values of c during computation,
    // it can not do the right alpha * a * b + beta * c operation.
    // Thus, the dpcpp kernel only adds alpha * a * b when it uses atomic
    // operation.
    if (atomic) {
```

Reply:
I will move this part into the atomic scope and the following is what I use (including c be scaled)
```
// Because the atomic operation changes the values of c during
// computation, it can not directly do alpha * a * b + beta * c
// operation. The beta * c needs to be done before calling this kernel.
// Then, this kernel only adds alpha * a * b when it uses atomic
// operation.
```